In [102]:
from PIL import Image
import pytesseract
import numpy as np

# Update the path if tesseract is not installed in the default path
# pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'

def image_to_array(image_path):
    # Open the image
    img = Image.open(image_path)
    
    # Use pytesseract to do OCR on the image
    text = pytesseract.image_to_string(img)
    
    # Convert the text into an array of lines or words
    # Split by lines to get an array of each line in the text
    lines_array = text.splitlines()
    
    # Convert the array to a numpy array if desired
    lines_array = np.array(lines_array)
    
    return lines_array

# Usage example
image_path = 'test/front.png'
front_text_array = image_to_array(image_path)
print(front_text_array[[8]])



['5730 2300 6317']


In [103]:
front_text_array

array(['SER Fears wr TAT B, ATT aT eae aT ET',
       'Feo Goa wears (sees Taree, UT YS HS /',
       'aime wRacace ht ean) a ares fae ora artes |', '',
       'Aadhaar is proof of identity, not of citizenship',
       'or date of birth. It should be used with verification (online',
       'authentication, or scanning of QR code / offline XML).', '',
       '5730 2300 6317', 'ART arene, ALT Geadiet', ''], dtype='<U61')

In [ ]:
len(str(front_text_array[[12]]))

18

In [81]:
from PIL import Image
import pytesseract
import numpy as np

# Update the path if tesseract is not installed in the default path
# pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'

def image_to_array(image_path):
    # Open the image
    img = Image.open(image_path)
    
    # Use pytesseract to do OCR on the image
    text = pytesseract.image_to_string(img)
    
    # Convert the text into an array of lines or words
    # Split by lines to get an array of each line in the text
    lines_array = text.splitlines()
    
    # Convert the array to a numpy array if desired
    lines_array = np.array(lines_array)
    
    return lines_array

# Usage example
image_path = 'test/back.png'
back_text_array = image_to_array(image_path)
print(back_text_array)



['Ua:' '' 'AGE seh, OT -9, TS aL, Gage sie' '' 'gars - 831012' '=Address:'
 'SMaqsood Ali, H NO -1, ROAD NO 18, J KS' 'SCOLONY JAWAHARNAGAR MANGO'
 'JAMSHEDPUR, Jamshedpur, PO: Mango,' '¥bist- East Singhbhum,'
 '© Jharkhand - 831012' '' '(=' '' '5730 2300 6317'
 'VID : 9140 2246 3221 3628' ''
 'Ge ics7 | 4 help@uidal.gov.in | GD www.uidal.gov.in']


In [82]:
text

'Ua:  AGE seh, OT -9, TS aL, Gage sie  gars - 831012 =Address: SMaqsood Ali, H NO -1, ROAD NO 18, J KS SCOLONY JAWAHARNAGAR MANGO JAMSHEDPUR, Jamshedpur, PO: Mango, ¥bist- East Singhbhum, © Jharkhand - 831012  (=  5730 2300 6317 VID : 9140 2246 3221 3628  Ge ics7 | 4 help@uidal.gov.in | GD www.uidal.gov.in'

In [69]:
ntext=back_text_array[6:14]
print(ntext)


['SMaqsood Ali, H NO -1, ROAD NO 18, J KS' 'SCOLONY JAWAHARNAGAR MANGO'
 'JAMSHEDPUR, Jamshedpur, PO: Mango,' '¥bist- East Singhbhum,'
 '© Jharkhand - 831012' '' '(=' '']


In [70]:
ctext = ' '.join(ntext)
print(ctext)

SMaqsood Ali, H NO -1, ROAD NO 18, J KS SCOLONY JAWAHARNAGAR MANGO JAMSHEDPUR, Jamshedpur, PO: Mango, ¥bist- East Singhbhum, © Jharkhand - 831012  (= 


In [80]:
import re

# Join array elements into a single string
back_text_array_str = ' '.join(back_text_array)

# Regular expression with capture groups
pattern = r'Address(.*?)(\b\w+\b)(.*?)831012'

# Capture all text except the first word after address and special characters
match = re.search(pattern, text)

if match:
  # Group 1: Text before the first word after address
  address_text = match.group(1)
  # Group 3: Text after the first word (excluding special characters)
  main_text = re.sub(r"[^\w\s]", "", match.group(3))

  # Combine and print the desired text
  print(main_text)
else:
  print("Pattern not found.")
  

 Ali H NO 1 ROAD NO 18 J KS SCOLONY JAWAHARNAGAR MANGO JAMSHEDPUR Jamshedpur PO Mango bist East Singhbhum  Jharkhand  


In [65]:
main_text

' Ali H NO 1 ROAD NO 18 J KS SCOLONY JAWAHARNAGAR MANGO JAMSHEDPUR Jamshedpur PO Mango bist East Singhbhum  Jharkhand  '

In [71]:
ctext

'SMaqsood Ali, H NO -1, ROAD NO 18, J KS SCOLONY JAWAHARNAGAR MANGO JAMSHEDPUR, Jamshedpur, PO: Mango, ¥bist- East Singhbhum, © Jharkhand - 831012  (= '

In [72]:
import Levenshtein
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

def levenshtein_similarity(str1, str2):
    distance = Levenshtein.distance(str1, str2)
    max_len = max(len(str1), len(str2))
    similarity = 1 - distance / max_len
    return similarity

def jaccard_similarity(str1, str2):
    vectorizer = CountVectorizer()
    vectors = vectorizer.fit_transform([str1, str2])
    similarity = cosine_similarity(vectors)[0, 1]
    return similarity

# Example usage:

levenshtein_sim = levenshtein_similarity(ctext, main_text)
jaccard_sim = jaccard_similarity(ctext, main_text)

print("Levenshtein Similarity:", levenshtein_sim)
print("Jaccard Similarity:", jaccard_sim)

Levenshtein Similarity: 0.7866666666666666
Jaccard Similarity: 0.9591663046625443


Levenshtein Distance:

Calculates the minimum number of edits (insertions, deletions, substitutions) required to transform one string into another.
A lower distance indicates higher similarity.
The similarity score is calculated as 1 - (distance / max_length).
Jaccard Similarity:

Converts strings into numerical representations using a vectorizer.
Calculates the cosine similarity between the two vectors, which measures the cosine of the angle between them.
A higher cosine similarity indicates higher similarity.
Choosing the Right Metric:

Levenshtein Distance: Suitable for comparing strings with minor differences, such as typos or small variations.
Jaccard Similarity: Better for comparing strings with semantic similarity, where the meaning of the words is more important than the exact word order.

A Levenshtein similarity score of 0.7866666666666666 indicates that the two strings are relatively similar. This means that a relatively small number of edits (insertions, deletions, or substitutions) are required to transform one string into the other.

Here's a breakdown of how the score is calculated:

Levenshtein Distance:

This measures the minimum number of edits needed to transform one string into another.
A lower distance indicates higher similarity.
Similarity Calculation:

The similarity score is calculated using the formula:
similarity = 1 - (distance / max_len)
distance is the Levenshtein distance.
max_len is the maximum length of the two strings.
A higher similarity score, like 0.7866666666666666, suggests that the two strings have a substantial overlap in their characters and overall structure.

Example:

Consider these two strings:

String 1: "hello world"
String 2: "hello world!"
The Levenshtein distance between these two strings is 1, as only one edit (adding an exclamation point) is required to transform String 1 into String 2.

from re import T
import active_liveness
import streamlit as st
# from streamlit import legacy_caching
from streamlit_webrtc import webrtc_streamer, VideoProcessorBase, ClientSettings
import cv2
import av
import numpy as np
from ocr.back_side_parse import get_info_back, get_string_similarity
import datetime

import os, sys
from active_liveness import headpose_liveness

my_absolute_dirpath = os.path.abspath(os.path.dirname(__file__))
sys.path.insert(0, my_absolute_dirpath + '/passive_liveness')
from passive_liveness import predict_liveness

sys.path.insert(0, my_absolute_dirpath + '/face')
from face import dnn_face_detection, find_face_similarity

import Levenshtein
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

import re
from PIL import Image
import pytesseract
import numpy as np

from re import T
import active_liveness
import streamlit as st
# from streamlit import legacy_caching
from streamlit_webrtc import webrtc_streamer, VideoProcessorBase, ClientSettings
import cv2
import av
import numpy as np
from ocr.back_side_parse import get_info_back, get_string_similarity
import datetime

import os, sys
from active_liveness import headpose_liveness

my_absolute_dirpath = os.path.abspath(os.path.dirname(__file__))
sys.path.insert(0, my_absolute_dirpath + '/passive_liveness')
from passive_liveness import predict_liveness

sys.path.insert(0, my_absolute_dirpath + '/face')
from face import dnn_face_detection, find_face_similarity

import Levenshtein
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

import re
from PIL import Image
import pytesseract
import numpy as np

In [84]:
from re import T
import active_liveness
import streamlit as st
# from streamlit import legacy_caching
from streamlit_webrtc import webrtc_streamer, VideoProcessorBase, ClientSettings
import cv2
import av
import numpy as np
from ocr.back_side_parse import get_info_back, get_string_similarity
import datetime

import os, sys

import Levenshtein
from sklearn.metrics.pairwise import cosine_similarity

import re
from PIL import Image
import pytesseract
import numpy as np


In [104]:
import cv2
citizenship_front = cv2.imread("test/front.png")
citizenship_back = cv2.imread("test/back.png")


In [127]:
import cv2
import numpy as np
import matplotlib.pyplot as plt

image1_path = 'test/front.png'
image2_path = 'test/back.png'

img1 = cv2.imread(image1_path)
if img1 is None:
    print("Error: Could not load image:", image1_path)
    exit()

img2 = cv2.imread(image2_path)
if img2 is None:
    print("Error: Could not load image:", image2_path)
    exit()

def get_info_back(image):
    """Returns all the information from input image

    Args:
        image (ndarray): input image

    Returns:
        dic: information from the input image
    """
    # os.environ["TESSDATA_PREFIX"] = r"C:\Program Files\Tesseract-OCR\tessdata"
    image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    text = pytesseract.image_to_string(image, lang='eng')
    text = [line for line in text.split('\n') if line.strip() != '']

    info = {}

    # cnText = [i.split(' ') for i in text]
    cnText = []
    for i in text:
        ct = i.split(' ')
        ct = [re.sub("[$&+,:;=?@#|'<>.^*()%!-]", '', c) for c in ct]
        ct = list(filter(None, ct))
        cnText.append(ct)
    
    return cnText


In [128]:
info = get_info_back(img1)


In [129]:
info

[['Aadhaar', 'no', 'issued', '24/10/2013'],
 ['ana', 'BHT'],
 ['Government', 'of', 'India'],
 ['e', 'ett'],
 ['Shaad', 'Ali'],
 ['a', 'fafei/DOB', '08/08/2003'],
 ['gee/', 'MALE'],
 ['Sea', 'ears', 'wr', 'TAT', 'arr', 'at', 'reafefer', 'a', 'eT'],
 ['wee', 'STENT', 'wears', 'site', 'Tae', 'aT', 'YSN', 'wrS/'],
 ['Simons', 'wracace', 'at', 'eft', 'a', 'ares', 'fen', 'ot', 'aTfey'],
 ['Aadhaar', 'is', 'proof', 'of', 'identity', 'not', 'of', 'citizenship'],
 ['or',
  'date',
  'of',
  'birth',
  'It',
  'should',
  'be',
  'used',
  'with',
  'verification',
  'online'],
 ['authentication',
  'or',
  'scanning',
  'of',
  'QR',
  'code',
  '/',
  'offline',
  'XML'],
 ['5730', '2300', '6317'],
 ['Art', 'aem', 'AN', 'veut']]

In [163]:
def find_target_list_regex(data, pattern):
    target_list = []
    for sublist in data:
        # Join the sublist into a string and search for the pattern
        joined_str = ' '.join(sublist)
        match = re.findall(pattern, joined_str)
        
        # If there are exactly 3 matches, add it to the result list
        if len(match) == 3:
            target_list.append(match)
    
    return target_list

In [165]:
target_pattern = r'\b\d{4}\b'

# Find the target list using the regex pattern
target_list = find_target_list_regex(info, target_pattern)

# Output the result
print(target_list)

[['5730', '2300', '6317']]


In [175]:
result = ' '.join(target_list[0])
# Output the result
print(result)

5730 2300 6317


In [177]:
len(result)

14

In [167]:
info1 = get_info_back(img2)


In [168]:
info1

[['bd', 'ade', 'fake', 'gear', 'wiitraror', '~'],
 ['Unique', 'Identification', 'Authority', 'of', 'India', 'ZAM'],
 ['Ua'],
 ['FORE', 'Het', 'Far', 'A', '9', 'Us', '9c', 'Sears', 'sie'],
 ['gerers', '831012'],
 ['Address'],
 ['SMaqsood', 'Ali', 'H', 'NO', '1', 'ROAD', 'NO', '18', 'KS'],
 ['COLONY', 'JAWAHARNAGAR', 'MANGO'],
 ['JAMSHEDPUR', 'Jamshedpur', 'PO', 'Mango'],
 ['3bist', 'East', 'Singhbhum'],
 ['©', 'Jharkhand', '831012'],
 ['5730', '2300', '6317'],
 ['VID', '9140', '2246', '3221', '3628'],
 ['Geics7_', '4', 'helpuidalgovin', 'GD', 'wwwuidalgovin']]

In [169]:


# Define the variable L (which will be set dynamically)
L = result   # Modify this as needed

# Flatten the list and join each inner list without spaces
flattened_list = [' '.join(sublist) for sublist in info1]

# Find the index of "Address" and L in the flattened list
address_index = flattened_list.index('Address')
try:
    L_index = flattened_list.index(L)
except ValueError:
    L_index = None  # Handle case where L is not found

# Extract elements between "Address" and L
if L_index and L_index > address_index:
    result1 = ''.join(flattened_list[address_index + 1 : L_index])  # Join elements between Address and L
else:
    result1 = ""  # If L is not found or not in the correct order

# Output the result
print(result1)


SMaqsood Ali H NO 1 ROAD NO 18 KSCOLONY JAWAHARNAGAR MANGOJAMSHEDPUR Jamshedpur PO Mango3bist East Singhbhum© Jharkhand 831012
